pkg/controller/hibernation/ibmcloud_actuator.go

```diff
@@ -0,0 +1,173 @@
+
+	return nil
+}
+
+func ibmCloudInstanceNames(instances []vpcv1.Instance) []string {
+	names := make([]string, len(instances))
+	for i, instance := range instances {
+		names[i] = *instance.Name
+	}
+	return names
+}
+
+// MachinesRunning will return true if the machines associated with the given
+// ClusterDeployment are in a running state. It also returns a list of machines that
+// are not running.
+func (a *ibmCloudActuator) MachinesRunning(cd *hivev1.ClusterDeployment, hiveClient client.Client, logger log.FieldLogger) (bool, []string, error) {
+	logger = logger.WithField("cloud", "ibmcloud")
+	logger.Infof("checking whether machines are running")
+	ibmCloudClient, err := a.ibmCloudClientFn(cd, hiveClient, logger)
+	if err != nil {
+		return false, nil, err
+	}
+	instances, err := getIBMCloudClusterInstances(cd, ibmCloudClient, notRunningStates, logger)
+	if err != nil {
+		return false, nil, err
+	}
+	return len(instances) == 0, ibmCloudInstanceNames(instances), nil
+}
+
+// MachinesStopped will return true if the machines associated with the given
+// ClusterDeployment are in a stopped state. It also returns a list of machines
+// that have not stopped.
+func (a *ibmCloudActuator) MachinesStopped(cd *hivev1.ClusterDeployment, hiveClient client.Client, logger log.FieldLogger) (bool, []string, error) {
+	logger = logger.WithField("cloud", "ibmcloud")
+	logger.Infof("checking whether machines are stopped")
+	ibmCloudClient, err := a.ibmCloudClientFn(cd, hiveClient, logger)
+	if err != nil {
+		return false, nil, err
+	}
+	instances, err := getIBMCloudClusterInstances(cd, ibmCloudClient, notStoppedStates, logger)
+	if err != nil {
+		return false, nil, err
+	}
+	return len(instances) == 0, ibmCloudInstanceNames(instances), nil
+}
+
+func getIBMCloudClient(cd *hivev1.ClusterDeployment, c client.Client, logger log.FieldLogger) (ibmclient.API, error) {
+	secret := &corev1.Secret{}
+	err := c.Get(context.TODO(), client.ObjectKey{Name: cd.Spec.Platform.IBMCloud.CredentialsSecretRef.Name, Namespace: cd.Namespace}, secret)
+	if err != nil {
+		logger.WithError(err).Log(controllerutils.LogLevel(err), "failed to fetch IBM Cloud credentials secret")
+		return nil, errors.Wrap(err, "failed to fetch IBM Cloud credentials secret")
+	}
+	return ibmclient.NewClientFromSecret(secret)
+}
+
+func getIBMCloudClusterInstances(cd *hivev1.ClusterDeployment, c ibmclient.API, states sets.String, logger log.FieldLogger) ([]vpcv1.Instance, error) {
+	infraID := cd.Spec.ClusterMetadata.InfraID
+	logger = logger.WithField("infraID", infraID)
+	logger.Debug("listing cluster instances")
+
+	// The ResourceGroup used here defaults to the Cluster InfraID.
+	// This is not the case BUT if our IBMCloud platform had a field for a user provided ResourceGroupName then we would have to use that ResourceGroupName here.
+	resourceGroup, err := c.GetResourceGroup(context.TODO(), infraID)
+	if err != nil {
+		logger.WithError(err).Errorf("failed to get resource group %q", infraID)
```

Comment:
Missed return here (please add UT). Also, structured logging.
```suggestion
		logger.WithError(err).WithField("resourceGroup", infraID).Error("failed to get resource group")
		return nil, err
```